### PR TITLE
provide some generic functions to reduce required ring interface

### PR DESCRIPTION
**Pull request by tthsqe12**
this was fun!

### CODE REVIEW

**Comment by thofma** [comment]
Nemo already defines `valuation(::Integer, ::Integer)` and `remove(::Integer, ::Integer)` by promoting everything to `fmpz`. :disappointed:

File: src/algorithms/GenericFunctions.jl [hunk 1]
```diff
@@ -4,6 +4,152 @@
 #
 ###############################################################################
 
+function ^(a::T, n::Integer) where T <: RingElem
+   n < 0 && throw(DomainError(n, "exponent must be >= 0"))
```
Comment (thofma) [inline src/algorithms/GenericFunctions.jl:8]: Maybe try `inv`? We are a bit inconsistent here. `fmpz(1)^-1` works, but `one(QQ["x"][1])` fails.

**Review by wbhart — APPROVED** [review]
Looks good to me. Of course we could go overboard with this, but what is there so far is very logical to me.

I intend to add some generic functions for square/nth roots at some point based on the existence of sqrt and root. It might also be possible to do some divexact derived stuff as well, but I need to think about that carefully.

**Comment by thofma** [comment]
There is a clash with Nemo, which also defines `remove` and `valuation` for julia integers in a different way (different return type). Maybe the two packages should just be merged again.

**Comment by wbhart** [comment]
I support merging Nemo and AbstractAlgebra again, though obviously we need to keep the original two packages available for the Julia community.

I would personally leave AA for the Julia community and merge it back into Nemo, which would not depend on it directly any more, but have it merged back in.

I'm a little busy right now to start working on such a project, and am also about to take a week of leave. But I'd be up for it later this year, once I've cleared a few of the tasks off my list that Claus has set.

I'd be really keen on having some very nice unified docs for Nemo/AA, and use m4, pandoc or even Julia to preprocess the docs and replace ${PACKAGE_NAME} or something like that, with either Nemo or Oscar depending where the docs are imported. We could have a script which goes through the docs and replaces any occurrence with Nemo or Oscar or AbstractAlgebra with ${PACKAGE_NAME} for the inevitable times that someone accidentally put one of those directly in the docs.

**Comment by thofma** [comment]
@tthsqe12 To fix it, you could change the signature of `remove` and `valutation` to `Union{RingElem, Number}` instead of `RingElement`. This should make the `Nemo` methods more specific fixing the ambiguities.

**Comment by wbhart** [comment]
By the way, one really good argument for merging AA and Nemo again is that I found a colleague overseas who didn't realise Nemo was still being maintained. When people look at the repo activity they see a few commits here and there wrapping this or that function/library, but no real work to extend the system or do generics or whatever.

Here is a quote from what he wrote (which I think contains no personal information, so should be fine to quote here):

"Is Nemo basically "putting generics on top of Flint"?  Did anything
end up doing that, or did you do that by autogenerating C code, or just not do it?"

He then went on to extoll the virtues of Flint and say how impressed he was with all the cool functionality in it and how much he values it for his own work.

I wonder how many other people have the same initial impression of Nemo (and Flint). We made a big initial splash with Nemo, a website for it and so on, and then it just died, from most people's impression. No one has any clue what we've been doing.

**Comment by tthsqe12** [comment]
Hold on - I noticed a slow down in one of the tests that I have to investigate. And of course the ambiguous nonsense.

File: src/algorithms/GenericFunctions.jl (hunk 1 above)
Comment (tthsqe12) [reply src/algorithms/GenericFunctions.jl:8]: One of the problems for ^ and powmod is that the unary - doesn't work on `typemin(Int)`. and then `unsigned(Int)` works but `unsigned(BigInt)` doesn't. :(

**Comment by tthsqe12** [comment]
To be fair, there is https://github.com/Nemocas/Nemo.jl/issues/1156 for the `remove(::Integer, ::Integer)`.

**Pull request merged**